onnxruntime/contrib_ops/rocm/bert/tunable_op.h

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+#include "contrib_ops/rocm/bert/timer.h"
+
+namespace onnxruntime {
+namespace contrib {
+namespace rocm {
+
+struct OpParams {
+  virtual std::string signature() const = 0;
+};
+
+class Op {
+ public:
+  Op() : repeats_(100) {}
+
+  virtual void Run(const OpParams*) = 0;
+
+  void SetRepeats(int n) {
+    repeats_ = n;
+  }
+
+  float Profile(const OpParams* op_params) {
+    // warm up
+    for (int i = 0; i < 5; i++) {
+      Run(op_params);
+    }
+    timer_.Start();
+    for (int i = 0; i < repeats_; i++) {
+      Run(op_params);
+    }
+    timer_.End();
+    return timer_.time()/repeats_;
+  }
+
+  virtual ~Op() {}
+
+ private:
+  Timer timer_;
+  int repeats_;
+};
+
+class TunableOp {
+ public:
+  TunableOp() {}
+
+  void Run(const OpParams* op_params_) {
+    int id;
+    if (kernel_map_.find(op_params_->signature()) == kernel_map_.end()) {
+      id = FindFastest(op_params_);
+      kernel_map_.insert({op_params_->signature(), id});
+    } else {
+      id = kernel_map_[op_params_->signature()];
+    }
+    ops_[id]->Run(op_params_);
+  }
+
+  virtual ~TunableOp() {
+    for (auto& op : ops_) {
+      delete op;
+    }
+  }
+
+ protected:
+  std::vector<Op*> ops_;
```

Comment:
Maybe use std::vector<std::shared_ptr<Op>> for this, so that the dtor can be avoided.

Reply:
Yeah, that's good point. Initially I used unique_ptr and wasn't able to get it work with the error below. I think it could be something to try again in future and in a separate PR.

/usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:1200:7: note: candidate function not viable: no known conversion from 'unique_ptr<VectorAddOp<__half, 64, 1>, default_delete<VectorAddOp<__half, 64, 1>>>' to 'unique_ptr<Op, default_delete<Op>>' for 1st argument
      push_back(value_type&& __x)